text/0024-schema-syntax.md

```diff
@@ -0,0 +1,396 @@
+        isPrivate: Boolean,
+        publicAccess: String,
+        viewACL: DocumentShare,
+        modifyACL: DocumentShare,
+        manageACL: DocumentShare
+    };
+    entity DocumentShare;
+    entity Public in [DocumentShare];
+    entity Drive;
+
+    action CreateDocument, CreateGroup 
+        appliesTo { principal: [User], resource: [Drive] };
+    action ViewDocument 
+        appliesTo { principal: [User,Public], resource: [Document] };
+    action DeleteDocument, ModifyDocument, EditIsPrivate, AddToShareACL, EditPublicAccess 
+        appliesTo { principal: [User], resource: [Document] };
+    action ModifyGroup, DeleteGroup 
+        appliesTo { principal: [User], resource: [Group] };
+}
+```
+
+### Mockup: GitHub Model
+
+Here is a version of the [GitHub](https://github.com/cedar-policy/cedar-examples/blob/main/cedar-example-use-cases/github_example/schema.json) schema.
+
+```
+namespace "" {
+    entity User in [UserGroup,Team];
+    entity UserGroup in [UserGroup];
+    entity Repository {
+        readers: UserGroup,
+        triagers: UserGroup,
+        writers: UserGroup,
+        maintainers: UserGroup,
+        admins: UserGroup   
+    };
+    entity Issue {
+        repo: Repository,
+        reporter: User
+    };
+    entity Org {
+        members: UserGroup,
+        owners: UserGroup,
+        memberOfTypes: UserGroup
+    };
+    action pull, fork, push, add_reader, add_triager, add_writer, add_maintainer, add_admin 
+        appliesTo { principal: [User], resource: [Repository] };
+    action delete_issue, edit_issue, assign_issue 
+        appliesTo { principal: [User], resource: [Issue] };
+}
+```
+
+### Grammar
+
+Here is a grammar for the proposed schema syntax.
+
+The grammar applies the following the conventions. Capitalized words stand for grammar productions, and lexical tokens are given in all-caps. When productions or tokens match those in the Cedar policy grammar, we use the same names (e.g., `IDENT` and `Path`).
+
+For grammar productions it uses `|` for alternatives, `[]` for optional productions, `()` for grouping, and `{}` for repetition of a form zero or more times. 
+
+Tokens are defined using regular expressions, where `[]` stands for character ranges; `|` stands for alternation; `*` , `+` , and `?` stand for zero or more, one or more, and zero or one occurrences, respectively; `~` stands for complement; and `-` stands for difference. The grammar ignores whitespace and comments.
+
+```
+Schema    := {Namespace}
+Namespace := 'namespace' '"' Path '"' '{' {Decl} '}'
+Decl      := Entity | Action | TypeDecl
+Entity    := 'entity' IDENT ['in' (EntType | '[' [EntTypes] ']')] [['='] RecType] ';'
+Action    := 'action' Name ['in' (Name | '[' [Names] ']')] [AppliesTo] ';'
+TypeDecl  := 'type' IDENT '=' Type ';'
+Type      := PRIMTYPE | IDENT | SetType | RecType
+EntType   := Path
+SetType   := 'Set' '<' Type '>'
+RecType   := '{' [AttrDecls] '}'
+AttrDecls := Name ':' Type [',' | ',' AttrDecls]
```

Comment:
In Schema-based format, record attributes have `required` fields. They are missing here.

Reply:
Good point! I will modify this to add an optional `?` after the name, in the style of Typescript. So the grammar should be
```
AttrDecls := Name ['?'] ':' Type [',' | ',' AttrDecls]
```